### ISSUE TITLE
Value of `this` in top-level module function is undefined

### ISSUE THREAD

**Issue by mwpenny** [laverdet/isolated-vm#175]
According to MDN, `this` should evaluate to the global object when it is not set (unless strict mode is being used, in which case it should evaluate to `undefined`).

The following test code demonstrates that a `Context`'s global scope is _not_ used for `this` when a `Module` is evaluated.

```js
const ivm = require('isolated-vm');

let isolate = new ivm.Isolate();
let context = isolate.createContextSync();

context.global.setSync('global', context.global.derefInto());
context.global.setSync('test', 'pass');

const code = `
    (function() {
        // Run when mod.evaluateSync() is called. Should evaluate to the global object
        const topLevelThis = this;

        global.main = function() {
            if (typeof topLevelThis === 'undefined') {
                throw new Error('top level this was undefined');
            } else if (topLevelThis.test !== 'pass') {
                throw new Error('test property was not found on top level this');
            }
        }
    })();
`;

const mod = isolate.compileModuleSync(code);
mod.instantiateSync(context, () => {} /* no dependencies */)
mod.evaluateSync();

const main = context.global.getSync('main', { reference: true });
main.applySync(null, []);
console.log('pass');
```

This problem does not occur if I use `Script.runSync()`. Am I misunderstanding the semantics of modules?

**Comment by laverdet** [comment]
Modules always have strict mode enabled, it's in the specification. 

You can confirm via plain nodejs:
```
$ node --input-type module -e 'console.log(this)'
undefined

$ node --input-type module -e 'with({});'
file:///Users/marcel/[eval1]:1
with({});
^^^^
SyntaxError: Strict mode code may not include a with statement
```

You can use `globalThis` which was recently to the JS spec and works in nodejs v12+. It takes the role of unifying `window`, `this`, `self` and `global`.

**Comment by mwpenny** [comment]
I'm not talking about top-level `this`, I'm talking about `this` inside of an IIFE at the top-level. Such as:

```js
// module.js

(function() {
    console.log(this);
})();

module.exports = "hello";
```

When I run `require('./module')` from within node, `global` is printed.

**Comment by mwpenny** [comment]
This makes it seem like strict mode is not the default for node modules. Although you are correct, according to https://v8.dev/features/modules, strict mode is default. I suppose it is not possible to pass a flag to enable/disable it?

In any case, thank you for the clarification.

**Comment by laverdet** [comment]
node modules are by default [commonjs](https://en.wikipedia.org/wiki/CommonJS) modules, not [ecmascript modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import). Take a look at the [nodejs modules](https://nodejs.org/api/modules.html) documentation. node modules are parsed as javascript script files unless the nearest package.json has `"type":"module"` OR the extension ends in `.mjs`. If the code you're trying to run depends on non strict mode then it is most assuredly *not* a javascript module and should be compiled with `compileScript` instead of `compileModule`.

**Comment by mwpenny** [comment]
This all makes a lot more sense now. I should have known that in the JavaScript ecosystem of all places, "module" is a very overloaded term :)

I think I have it straight now. So in short, V8 modules are ECMAScript modules which have different rules than CommonJS/"classic" Node modules. The `Module` class in `isolated-vm` represents the former.

I was already providing `exports`/`module.exports` in the context's global scope and using it to grab the user code. So, like you said I ended up just switching `compileModule()` to `compileScript()` after I realized what I was compiling wasn't actually a native JS module and that solved my problem. I thought I had to use a `Module` for the user "module".

Thank you for this information, I've definitely learned something useful!
